### ISSUE TITLE
MultiHeadAttention layer save and load changes weight

### ISSUE THREAD

**Issue by yesufeng** [tensorflow/tensorflow#52764]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1.  It must be a bug, a feature request, or a significant problem with the
    documentation (for small docs fixes please send a PR instead).
2.  The form below must be filled out.
3.  It shouldn't be a TensorBoard issue. Those go
    [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**: Yes
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
-   **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue
    happens on a mobile device**:
-   **TensorFlow installed from (source or binary)**: pip install tensorflow
-   **TensorFlow version (use command below)**: v2.4.0-49-g85c8b2a817f 2.4.1
-   **Python version**: 3.7.11
-   **Bazel version (if compiling from source)**:
-   **GCC/Compiler version (if compiling from source)**:
-   **CUDA/cuDNN version**: https://aws.amazon.com/ec2/instance-types/g4/
-   **GPU model and memory**: https://aws.amazon.com/ec2/instance-types/g4/
-   **Exact command to reproduce**: see below in the source code session.

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:

```bash
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"
```

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

trained a small model with a MHA layer and save model and load it back, the weight of the MHA layer changed. I have checked similar thread like [this](https://github.com/tensorflow/tensorflow/issues/47722), the suggestion was to fix the seed given to the `kernel_initializer`, so did that as shown below, but does not fix the problem. Notably, if just build the model and save without training (`.fit`), loading back the model, then the outputs are consistent. 

### Source code

from tensorflow.keras.layers import MultiHeadAttention, Concatenate, Add, Dense
from tensorflow.keras.initializers import glorot_uniform
import tensorflow as tf
from tensorflow.keras.optimizers import Adam

    def encoder_layer_simple(units, d_model, num_heads, dropout, name="encoder_layer"):
        assert d_model % num_heads == 0
    
        depth = d_model // num_heads
        query = tf.keras.Input(shape=(None, d_model), name="query")
        key = tf.keras.Input(shape=(None, d_model), name="key")
        value = tf.keras.Input(shape=(None, d_model), name='value')
    
        attention, attn_weights = MultiHeadAttention(num_heads, key_dim=depth, value_dim=depth, dropout=dropout,
                                                     kernel_initializer=glorot_uniform(seed=0))(
            query=query, value=value, key=key, return_attention_scores=True)
        attention = tf.keras.layers.LayerNormalization(epsilon=1e-6)(attention + value)
    
        # ffwd sub-layer
        outputs = tf.keras.layers.Dense(units=units, activation='relu')(attention)
        outputs = tf.keras.layers.Dense(units=d_model)(outputs)
        outputs = tf.keras.layers.Dropout(rate=dropout)(outputs)
    
        outputs = tf.keras.layers.LayerNormalization(epsilon=1e-6)(outputs + attention)
    
        return tf.keras.Model(inputs=[query, key, value], outputs=outputs, name=name)

#### then create an 'encoder' model and do a fit
    encoder = encoder_layer_simple(512, 400, 8, 0, name='test_encoder')
    
    query = tf.random.uniform([256, 100, 400])
    
    enc_x = {'query':query, 'value':query, 'key':query}
    enc_y = tf.random.uniform([256, 100, 400])
    encoder.compile(Adam(),'MSE')
    # if we don't do fit, then the results are consistent
    encoder.fit(x=enc_x, y=enc_y)
    
    save_path = "path_to_model"
    encoder.save(save_path)
    loaded = tf.keras.models.load_model(save_path, compile=False)
    
    eout = encoder(enc_x)
    lout = loaded(enc_x)
    
    np.allclose(lout.numpy(), eout.numpy()) # will output False

#### also can get weight matrix from the MHA layer, close but different

    encoder.get_layer("multi_head_attention").get_weights()[0][1, 4, 1:10]

array([ 0.01589977,  0.01451083, -0.00762749, -0.00915873,  0.00802898,
        0.00830225, -0.01598875,  0.01182209,  0.01533959], dtype=float32)
    
    loaded.get_layer("multi_head_attention").get_weights()[0][1, 4, 1:10]

array([ 0.01591637,  0.01452216, -0.00764309, -0.00916362,  0.0079951 ,
        0.0082739 , -0.01598743,  0.01181626,  0.01530811], dtype=float32)
